extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BuildTimeProfileProcessor.java

```diff
@@ -0,0 +1,115 @@
+package io.quarkus.arc.deployment;
+
+import static io.quarkus.arc.processor.Annotations.find;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationTarget;
+import org.jboss.jandex.AnnotationValue;
+import org.jboss.jandex.DotName;
+
+import io.quarkus.arc.WhenBuildTimeProfile;
+import io.quarkus.arc.WhenNotBuildTimeProfile;
+import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.arc.processor.Transformation;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.runtime.configuration.ProfileManager;
+
+public class BuildTimeProfileProcessor {
+
+    private static final DotName WHEN_BUILD_TIME_PROFILE = DotName.createSimple(WhenBuildTimeProfile.class.getName());
+    private static final DotName WHEN_NOT_BUILD_TIME_PROFILE = DotName.createSimple(WhenNotBuildTimeProfile.class.getName());
+
+    /**
+     * Uses {@link AnnotationsTransformer} to do the following:
+     *
+     * Adds the {@code @Alternative} annotation to any class or producer that is annotated with {code @WhenBuildTimeProfile}
+     * when the specified profile doesn't match the current one.
+     *
+     * Adds the {@code @AlternativePriority} annotation to any class or producer that is annotated with
+     * {code @WhenBuildTimeProfile} when the specified profile does match the current one.
+     *
+     * This effectively means that a bean annotated with {@code @WhenBuildTimeProfile} is only enabled when
+     * the current profile matches that of the annotation
+     */
+    @BuildStep
+    void whenBuildTimeProfile(BeanArchiveIndexBuildItem beanArchiveIndex,
+            CustomScopeAnnotationsBuildItem scopes,
+            BuildProducer<AnnotationsTransformerBuildItem> annotationsTransformer) {
+        annotationsTransformer.produce(new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {
+            @Override
+            public boolean appliesTo(AnnotationTarget.Kind kind) {
+                return kind == AnnotationTarget.Kind.METHOD || kind == AnnotationTarget.Kind.CLASS
+                        || kind == AnnotationTarget.Kind.FIELD;
+            }
+
+            @Override
+            public void transform(TransformationContext ctx) {
+                AnnotationInstance whenBuildTimeProfileInstance = find(ctx.getAnnotations(), WHEN_BUILD_TIME_PROFILE);
+                if (whenBuildTimeProfileInstance == null) {
+                    return;
+                }
+
+                Transformation transform = ctx.transform();
+
+                String profileOnInstance = whenBuildTimeProfileInstance.value().asString();
+                if (profileOnInstance.equals(ProfileManager.getActiveProfile())) {
+                    transform.add(DotNames.ALTERNATIVE_PRIORITY,
```

Comment:
Recently we added the ability to enable alternatives via config file and those also use `Integer.MAX_VALUE`.
So if a user has an alternative bean that is enabled that way and then adds another bean just for one given profile, he will actually end up with two alternatives with maximum priority (which I think should blow up). Not sure if this corner case is something we want to address right away, just writing it down :)

Reply:
Hm, yes that would blow up with ambiguous dependency. But I'd say that it's more like a corner case and maybe we could add a note to the docs? Or even use `Intereg.MAX_VALUE - 1` ;-).

Reply:
How about goign with `Intereg.MAX_VALUE - 1` for now and revisiting if this every becomes a problem?
Seems like an extreme corner case to me.

Reply:
+1 to `Intereg.MAX_VALUE - 1`, it just occured to me so I thought I'd mention it :)